csrc/host_ir/executor.cpp

```diff
@@ -325,24 +325,38 @@ void HostIrEvaluator::handle(Synchronize* synchronize) {
 }
 
 void HostIrEvaluator::handle(LaunchKernel* launch_kernel) {
-  KernelArgumentHolder args;
+  KernelArgumentHolder args, outputs;
+  bool preallocated_outputs = true;
   for (auto& input : launch_kernel->inputs()) {
     args.push(getKnownConcreteValue(input));
   }
+
+  // If all output buffers are known already, pass them to the executor
+  for (auto& output : launch_kernel->outputs()) {
+    if (expr_evaluator_.isKnown(output)) {
```

Comment:
When do you expect kernel outputs to be unknown? I expect all kernel outputs to be preallocated eventually. Maybe you are doing this just as an intermediate state before you also change compileFusionParallel?

Reply:
+1, we should rather assert that the outputs are known here, right?

Reply:
The outputs get allocated in KernelExecutor::run, so I'm expecting the outputs to always be unknown in this PR. Once I update compileFusionParallel this will change though since they'll be preallocated already.